sam.c

```diff
@@ -352,6 +352,33 @@ int32_t bam_endpos(const bam1_t *b)
         return b->core.pos + 1;
 }
 
+static void bam_tag2cigar(bam1_t *b)
+{
+    bam1_core_t *c = &b->core;
+    uint32_t cigar_st, n_cigar4, CG_st, CG_en, ori_len = b->l_data;
+    uint8_t *CG;
+    if (c->n_cigar > 0 || !(c->flag & BAM_FUNMAP) || c->tid < 0 || c->pos < 0) return;
+    if ((CG = bam_aux_get(b, "CG")) == 0) return;
+    if (CG[0] != 'B' || CG[1] != 'I') return;
+    cigar_st = (uint8_t*)bam_get_cigar(b) - b->data;
+    c->n_cigar = *(uint32_t*)(CG + 2);
+    n_cigar4 = c->n_cigar * 4;
+    CG_st = CG - b->data - 2;
+    CG_en = CG_st + 8 + n_cigar4;
+    b->l_data += n_cigar4;
+    if (b->m_data < b->l_data) {
+        b->m_data = b->l_data;
+        kroundup32(b->m_data);
+        b->data = (uint8_t*)realloc(b->data, b->m_data);
+    }
```

Comment:
Good point on realloc.  If we're in the usual pipeline of reusing the same bam then it'll not realloc very often so thumbs up on that idea.

Regarding truncated array, my worry is about malicious data where a malformed BAM file has been loaded.  My understanding is that the BAM reader just reads block_size bytes, but the contents of the aux data may not be consistent with the decoded block_size value.

It may not actually be a buffer write overrun though as by reallocing we're growing the array first before doing the memcpy, but it may lead to exposing uninitialised data.  Consider a broken aux field of CGBI<0xff 0x00 0x00 0x00>foo <end of stream>.  It claims to be 255 ints long, so nearly 1Kb, but only contains "foo" bytes.  The BAM record size itself could be correct, ending at the "foo" point, but the actual decoding of the BAM record would expose an error.  This code however will grow the BAM record by ~1Kb to make way for 255 cigar elements and then memcpy ~1Kb from ```CGBI``` onwards to that cigar array, which possibly includes nearly 1Kb of random contents from a previous malloc/free on the heap.

If bam_aux_append has the same issue then it needs fixing too, but on a quick glance it seems fine to me.